File: scripts/remove-lombok.py
#!/usr/bin/env python3
"""Remove Lombok annotations and generate explicit Java accessors."""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]

LOMBOK_IMPORT = re.compile(r"^import\s+lombok(?:\.[\w*]+|\.\*);\s*\n", re.MULTILINE)
LOMBOK_ANNOTATION = re.compile(
    r"^@(?:Data|Getter|Setter|Builder|Slf4j|RequiredArgsConstructor|NoArgsConstructor|AllArgsConstructor|EqualsAndHashCode|ToString|Value)(?:\([^)]*\))?\s*\n",
    re.MULTILINE,
)
FIELD_LINE = re.compile(
    r"^(?P<indent>\s*)(?P<mods>(?:@\w+(?:\([^)]*\))?\s+)*)"
    r"(?P<access>private|protected)\s+(?P<static>static\s+)?(?P<final>final\s+)?"
    r"(?P<type>[\w<>,\[\]?.@\s]+?)\s+(?P<name>\w+)\s*(?P<init>=\s*[^;]+)?;",
    re.MULTILINE,
)
EXISTING_METHOD = re.compile(
    r"^\s*(?:public|protected|private)\s+(?:static\s+)?(?:[\w<>,\[\]?.@\s]+\s+)?"
    r"(?P<name>(?:get|set|is)[A-Z]\w*)\s*\(",
    re.MULTILINE,
)
CLASS_DECL = re.compile(
    r"(?P<prefix>^[^{]*?\b(?P<kind>class|enum|record|interface)\s+(?P<name>\w+)[^{]*\{)",
    re.MULTILINE | re.DOTALL,
)


def capitalize(name: str) -> str:
    return name[0].upper() + name[1:] if name else name


def getter_name(field_name: str, field_type: str) -> str:
    if field_type == "boolean" or field_type == "Boolean":
        return f"is{capitalize(field_name)}"
    return f"get{capitalize(field_name)}"


def existing_accessors(content: str) -> set[str]:
    return {match.group("name") for match in EXISTING_METHOD.finditer(content)}


def parse_fields(body: str) -> list[dict]:
    fields: list[dict] = []
    for match in FIELD_LINE.finditer(body):
        if match.group("static"):
            continue
        field_type = " ".join(match.group("type").split())
        fields.append(
            {
                "name": match.group("name"),
                "type": field_type,
                "final": bool(match.group("final")),
                "indent": match.group("indent"),
            }
        )
    return fields


def find_class_body(content: str) -> tuple[str, int, int] | None:
    match = CLASS_DECL.search(content)
    if not match:
        return None
    start = match.end() - 1
    depth = 0
    for index in range(start, len(content)):
        char = content[index]
        if char == "{":
            depth += 1
        elif char == "}":
            depth -= 1
            if depth == 0:
                return match.group("name"), start + 1, index
    return None


def detect_lombok_flags(content: str) -> dict[str, bool]:
    return {
        "data": "@Data" in content,
        "getter": "@Getter" in content,
        "setter": "@Setter" in content,
        "no_args": "@NoArgsConstructor" in content,
        "all_args": "@AllArgsConstructor" in content,
        "required_args": "@RequiredArgsConstructor" in content,
        "slf4j": "@Slf4j" in content,
        "is_enum": bool(re.search(r"\benum\s+\w+", content)),
    }


def should_getter(flags: dict[str, bool]) -> bool:
    return flags["data"] or flags["getter"] or flags["is_enum"]


def should_setter(flags: dict[str, bool]) -> bool:
    return flags["data"] or flags["setter"]


def generate_getter(field: dict) -> str:
    name = getter_name(field["name"], field["type"])
    return (
        f"\n    public {field['type']} {name}() {{\n"
        f"        return {field['name']};\n"
        f"    }}\n"
    )


def generate_setter(field: dict) -> str:
    param = field["name"]
    method = f"set{capitalize(field['name'])}"
    return (
        f"\n    public void {method}({field['type']} {param}) {{\n"
        f"        this.{field['name']} = {param};\n"
        f"    }}\n"
    )


def generate_no_args_constructor(class_name: str, fields: list[dict], existing: set[str]) -> str:
    if "NoArgsConstructor" in existing or any(
        re.search(rf"public\s+{class_name}\s*\(\s*\)", line)
        for line in existing
    ):
        return ""
    return f"\n    public {class_name}() {{\n    }}\n"


def generate_all_args_constructor(class_name: str, fields: list[dict]) -> str:
    if not fields:
        return ""
    params = ", ".join(f"{field['type']} {field['name']}" for field in fields)
    assigns = "\n".join(f"        this.{field['name']} = {field['name']};" for field in fields)
    return (
        f"\n    public {class_name}({params}) {{\n"
        f"{assigns}\n"
        f"    }}\n"
    )


def generate_required_args_constructor(class_name: str, fields: list[dict]) -> str:
    required = [field for field in fields if field["final"]]
    if not required:
        return ""
    params = ", ".join(f"{field['type']} {field['name']}" for field in required)
    assigns = "\n".join(f"        this.{field['name']} = {field['name']};" for field in required)
    return (
        f"\n    public {class_name}({params}) {{\n"
        f"{assigns}\n"
        f"    }}\n"
    )


def generate_logger(class_name: str) -> tuple[str, str]:
    import_line = "import org.slf4j.Logger;\nimport org.slf4j.LoggerFactory;\n"
    field = f"\n    private static final Logger log = LoggerFactory.getLogger({class_name}.class);\n"
    return import_line, field


def transform_file(path: Path) -> bool:
    original = path.read_text(encoding="utf-8")
    if "lombok" not in original:
        return False

    flags = detect_lombok_flags(original)
    existing = existing_accessors(original)
    class_info = find_class_body(original)
    if not class_info:
        print(f"WARN: could not parse class body: {path}", file=sys.stderr)
        return False

    class_name, body_start, body_end = class_info
    body = original[body_start:body_end]
    fields = parse_fields(body)

    content = LOMBOK_IMPORT.sub("", original)
    content = LOMBOK_ANNOTATION.sub("", content)

    class_info = find_class_body(content)
    if not class_info:
        return False
    _, body_start, body_end = class_info

    generated: list[str] = []

    if flags["slf4j"]:
        import_line, logger_field = generate_logger(class_name)
        if "LoggerFactory" not in content:
            package_match = re.search(r"^package .+;\n", content, re.MULTILINE)
            if package_match:
                insert_at = package_match.end()
                content = content[:insert_at] + "\n" + import_line + content[insert_at:]
                class_info = find_class_body(content)
                if class_info:
                    _, body_start, body_end = class_info
        generated.append(logger_field)

    if flags["no_args"]:
        generated.append(generate_no_args_constructor(class_name, fields, existing))
    if flags["all_args"]:
        generated.append(generate_all_args_constructor(class_name, fields))
    if flags["required_args"]:
        generated.append(generate_required_args_constructor(class_name, fields))

    if should_getter(flags):
        for field in fields:
            getter = getter_name(field["name"], field["type"])
            if getter not in existing:
                generated.append(generate_getter(field))
                existing.add(getter)

    if should_setter(flags):
        for field in fields:
            if field["final"]:
                continue
            setter = f"set{capitalize(field['name'])}"
            if setter not in existing:
                generated.append(generate_setter(field))
                existing.add(setter)

    if not generated:
        path.write_text(content, encoding="utf-8")
        return True

    insertion = "".join(generated)
    updated = content[:body_end] + insertion + content[body_end:]
    path.write_text(updated, encoding="utf-8")
    return True


def main() -> int:
    java_files = [
        path
        for path in ROOT.rglob("*.java")
        if "legacy" not in path.parts and "target" not in path.parts
    ]
    changed = 0
    for path in java_files:
        if transform_file(path):
            changed += 1
            print(path.relative_to(ROOT))
    print(f"Updated {changed} files")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/remove-lombok.js
#!/usr/bin/env node
/** Remove Lombok annotations and generate explicit Java accessors. */

const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");

const LOMBOK_IMPORT = /^import\s+lombok(?:\.[\w*]+|\.\*);\s*\n/gm;
const LOMBOK_ANNOTATION =
  /^@(?:Data|Getter|Setter|Builder|Slf4j|RequiredArgsConstructor|NoArgsConstructor|AllArgsConstructor|EqualsAndHashCode|ToString|Value)(?:\([^)]*\))?\s*\n/gm;
const FIELD_LINE =
  /^(\s*)((?:@\w+(?:\([^)]*\))?\s+)*)?(private|protected)\s+(?:(static)\s+)?(?:(final)\s+)?([\w<>,\[\]?.@\s]+?)\s+(\w+)\s*(=\s*[^;]+)?;/gm;
const EXISTING_METHOD =
  /^\s*(?:public|protected|private)\s+(?:static\s+)?(?:[\w<>,\[\]?.@\s]+\s+)?((?:get|set|is)[A-Z]\w*)\s*\(/gm;
const CLASS_DECL =
  /^[^{]*?\b(class|enum|record|interface)\s+(\w+)[^{]*\{/ms;

function walk(dir, files = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (entry.name === "target" || entry.name === "legacy" || entry.name === ".git") continue;
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) walk(full, files);
    else if (entry.name.endsWith(".java")) files.push(full);
  }
  return files;
}

function capitalize(name) {
  return name ? name[0].toUpperCase() + name.slice(1) : name;
}

function getterName(fieldName, fieldType) {
  const t = fieldType.trim();
  if (t === "boolean" || t === "Boolean") return `is${capitalize(fieldName)}`;
  return `get${capitalize(fieldName)}`;
}

function existingAccessors(content) {
  const names = new Set();
  let m;
  const re = new RegExp(EXISTING_METHOD.source, EXISTING_METHOD.flags);
  while ((m = re.exec(content))) names.add(m[1]);
  return names;
}

function parseFields(body) {
  const fields = [];
  let m;
  const re = new RegExp(FIELD_LINE.source, FIELD_LINE.flags);
  while ((m = re.exec(body))) {
    if (m[4]) continue; // static
    fields.push({
      name: m[7],
      type: m[6].replace(/\s+/g, " ").trim(),
      final: Boolean(m[5]),
      indent: m[1],
    });
  }
  return fields;
}

function findClassBody(content) {
  const match = content.match(CLASS_DECL);
  if (!match) return null;
  const start = match.index + match[0].length - 1;
  let depth = 0;
  for (let i = start; i < content.length; i++) {
    if (content[i] === "{") depth++;
    else if (content[i] === "}") {
      depth--;
      if (depth === 0) return { className: match[2], bodyStart: start + 1, bodyEnd: i };
    }
  }
  return null;
}

function detectFlags(content) {
  return {
    data: content.includes("@Data"),
    getter: content.includes("@Getter"),
    setter: content.includes("@Setter"),
    noArgs: content.includes("@NoArgsConstructor"),
    allArgs: content.includes("@AllArgsConstructor"),
    requiredArgs: content.includes("@RequiredArgsConstructor"),
    slf4j: content.includes("@Slf4j"),
    isEnum: /\benum\s+\w+/.test(content),
  };
}

function generateGetter(field) {
  const name = getterName(field.name, field.type);
  return `\n    public ${field.type} ${name}() {\n        return ${field.name};\n    }\n`;
}

function generateSetter(field) {
  const method = `set${capitalize(field.name)}`;
  return `\n    public void ${method}(${field.type} ${field.name}) {\n        this.${field.name} = ${field.name};\n    }\n`;
}

function generateNoArgsConstructor(className) {
  return `\n    public ${className}() {\n    }\n`;
}

function generateAllArgsConstructor(className, fields) {
  if (!fields.length) return "";
  const params = fields.map((f) => `${f.type} ${f.name}`).join(", ");
  const assigns = fields.map((f) => `        this.${f.name} = ${f.name};`).join("\n");
  return `\n    public ${className}(${params}) {\n${assigns}\n    }\n`;
}

function generateRequiredArgsConstructor(className, fields) {
  const required = fields.filter((f) => f.final);
  if (!required.length) return "";
  const params = required.map((f) => `${f.type} ${f.name}`).join(", ");
  const assigns = required.map((f) => `        this.${f.name} = ${f.name};`).join("\n");
  return `\n    public ${className}(${params}) {\n${assigns}\n    }\n`;
}

function hasNoArgsConstructor(content, className) {
  return new RegExp(`public\\s+${className}\\s*\\(\\s*\\)`).test(content);
}

function transformFile(filePath) {
  const original = fs.readFileSync(filePath, "utf8");
  if (!original.includes("lombok")) return false;

  const flags = detectFlags(original);
  const existing = existingAccessors(original);
  let classInfo = findClassBody(original);
  if (!classInfo) {
    console.warn(`WARN: could not parse: ${filePath}`);
    return false;
  }

  const fields = parseFields(original.slice(classInfo.bodyStart, classInfo.bodyEnd));

  let content = original.replace(LOMBOK_IMPORT, "").replace(LOMBOK_ANNOTATION, "");
  classInfo = findClassBody(content);
  if (!classInfo) return false;

  const generated = [];

  if (flags.slf4j) {
    if (!content.includes("LoggerFactory")) {
      const pkg = content.match(/^package .+;\n/m);
      if (pkg) {
        const imports = "import org.slf4j.Logger;\nimport org.slf4j.LoggerFactory;\n";
        content = content.slice(0, pkg.index + pkg[0].length) + "\n" + imports + content.slice(pkg.index + pkg[0].length);
        classInfo = findClassBody(content);
      }
    }
    generated.push(`\n    private static final Logger log = LoggerFactory.getLogger(${classInfo.className}.class);\n`);
  }

  if (flags.noArgs && !hasNoArgsConstructor(content, classInfo.className)) {
    generated.push(generateNoArgsConstructor(classInfo.className));
  }
  if (flags.allArgs) generated.push(generateAllArgsConstructor(classInfo.className, fields));
  if (flags.requiredArgs) generated.push(generateRequiredArgsConstructor(classInfo.className, fields));

  const wantGetter = flags.data || flags.getter || flags.isEnum;
  const wantSetter = flags.data || flags.setter;

  if (wantGetter) {
    for (const field of fields) {
      const getter = getterName(field.name, field.type);
      if (!existing.has(getter)) {
        generated.push(generateGetter(field));
        existing.add(getter);
      }
    }
  }

  if (wantSetter) {
    for (const field of fields) {
      if (field.final) continue;
      const setter = `set${capitalize(field.name)}`;
      if (!existing.has(setter)) {
        generated.push(generateSetter(field));
        existing.add(setter);
      }
    }
  }

  const updated = content.slice(0, classInfo.bodyEnd) + generated.join("") + content.slice(classInfo.bodyEnd);
  fs.writeFileSync(filePath, updated, "utf8");
  return true;
}

let changed = 0;
for (const file of walk(ROOT)) {
  if (transformFile(file)) {
    changed++;
    console.log(path.relative(ROOT, file));
  }
}
console.log(`Updated ${changed} files`);
